### DISCUSSION TITLE
Using "network_mode: host" to run on host network ?

### DISCUSSION THREAD

**Discussion by ArnoBen** [wg-easy/wg-easy#281] [Q&A] [answered]
Hello,
I wish to use my host network in order to make clients accessible to the host, so I tried the following docker-compose:

```
version: '3.3'
services:
    wg-easy:
        image: weejewel/wg-easy
        container_name: wg-easy
        environment:
            - WG_HOST=wg.myserver.com
            - PASSWORD=mypassword
        volumes:
            - ./wg-easy:/etc/wireguard
        # ports:
        #     - '51820:51820/udp'
        #     - '51821:51821/tcp'
        network_mode: host
        cap_add:
            - NET_ADMIN
            - SYS_MODULE
        # sysctls:
        #     - net.ipv4.conf.all.src_valid_mark=1
        #     - net.ipv4.ip_forward=1
        restart: unless-stopped

```

I also uncommented `net.ipv4.ip_forward=1` in `/etc/sysctl.conf`. Not sure what `net.ipv4.conf.all.src_valid_mark=1` does, maybe I need to add this line?

With this setup, I can connect a client to my vpn and my server sees it is connected, but it won't transmit data.

Is there a way to make this work ?

---------------------------
This is a bit unrelated but for additional information, now when I try to run the initial docker-compose
```
version: '3.3'
services:
    wg-easy:
        image: weejewel/wg-easy
        container_name: wg-easy
        environment:
            - WG_HOST=wg.myserver.com
            - PASSWORD=mypassword
        volumes:
            - ./wg-easy:/etc/wireguard
        ports:
            - '51820:51820/udp'
            - '51821:51821/tcp'
        cap_add:
            - NET_ADMIN
            - SYS_MODULE
        sysctls:
            - net.ipv4.conf.all.src_valid_mark=1
            - net.ipv4.ip_forward=1
        restart: unless-stopped
```

I get the following error:
```
Removing wg-easy
Recreating 74d032e20934_wg-easy ... 
Recreating 74d032e20934_wg-easy ... error

ERROR: for 74d032e20934_wg-easy  Cannot start service wg-easy: driver failed programming external connectivity on endpoint wg-easy (c041f02d986c8456407631a3eefda6dd2b0454742f32398509e00dd909691182): Error starting userland proxy: listen udp4 0.0.0.0:51820: bind: address already in use

ERROR: for wg-easy  Cannot start service wg-easy: driver failed programming external connectivity on endpoint wg-easy (c041f02d986c8456407631a3eefda6dd2b0454742f32398509e00dd909691182): Error starting userland proxy: listen udp4 0.0.0.0:51820: bind: address already in use
ERROR: Encountered errors while bringing up the project.
```

**Comment by shizunge** [comment]
You may need to update the post up script.

One of the default iptables settings in config.js is:
```
iptables -t nat -A POSTROUTING -s ${module.exports.WG_DEFAULT_ADDRESS.replace('x', '0')}/24 -o eth0 -j MASQUERADE
```

Where `eth0` should be the interface of your host network.

Of course you also need other iptables settings from the config.js.

**Comment by swallace17** [comment]
@shizunge I am also attempting to run this container with host networking, but for different reasons. I want to assign the container to a network interface on the host which is tagged to a vlan, such that this wireguard vpn's clients only have access to a particular subnet of my network. 

Reading through what you said, I believe you're saying that wg-easy container should be able to be run in host networking mode if you modify the `eth0` value in line 20 of /src/config.js to match whatever the host interface value is? 

I tested myself using the `ip link show` command inside my wg-easy container running with host networking. Instead of `eth0`, it looks like my primary interface is `eth0@if12`. 

So if I change line 20 of /src/config.js to the below and rebuild the container, my post up script should be correct, and I'm good to go? You mentioned needing other iptable settings from the same file. Could you expand on that? Thanks!

`iptables -t nat -A POSTROUTING -s ${module.exports.WG_DEFAULT_ADDRESS.replace('x', '0')}/24 -o eth0@if12 -j MASQUERADE`

*********************
Update: I had the chance to test. Inside my container, I ran the below command: 
`iptables -t nat -A POSTROUTING -s 10.8.0.0/24 -o eth0@if12 -j  MASQUERADE`. Connected clients are still not able to pass any traffic to the web, etc. I'm planning to change line 20 directly inside the container and rebuild, but have not had the time yet.

**Comment by shizunge** [comment]
I believe you should use 'eth0' rather than 'eth0@if12', though I am not an expert of Iptables.

I have not used the host network yet due to other limitation. I am sharing my experience where was going wrong last time I met the same problem that peer to peer connection was established, but client cannot access the Internet.

To my understanding (may not be 100% accurate) this is the meaning of the following Iptables

iptables -A FORWARD -i wg0 -j ACCEPT;
Allow forward traffic from the client to Internet.

iptables -t nat -A POSTROUTING -s ${module.exports.WG_DEFAULT_ADDRESS.replace('x', '0')}/24 -o eth0 -j MASQUERADE;
Record the original source of an internet request, when receiving the response, forward it back to the original client.

You also need the following or the forward won't work.
   cap_add:
      - NET_ADMIN
      - SYS_MODULE
    sysctls:
      - net.ipv4.ip_forward=1
      - net.ipv4.conf.all.src_valid_mark=1

I am not sure whether sysctls works with host network mode. You may try to add them to your host directly. Just my two cents.

Wireguard official website may have additional information and examples.

Good luck.

**Comment by shizunge** [comment]
Did you run

sudo sysctl -p

To reload the sysctl configuration after you modifies it on the host?

**Comment by swallace17** [comment]
@shizunge I did try running sudo sysctl -p after. Still no luck unfortunately. I'm not sure what is breaking in host mode that works fine in bridge. When you last attempted, it had  something to do with incorrect iptables?

**Comment by shizunge** [comment]
I am not using it in host mode yet. Sorry I cannot provide additional help. As I mentioned above  I was just sharing my experience where was going wrong last time I met the same problem that peer to peer connection was established, but client cannot access the Internet.

**Comment by swallace17** [comment]
@shizunge thats what I thought, no worries. Thanks anyway! In the meantime, I'm running this container inside a VM (whose network interface is tagged to a vlan). Works for now, but not ideal.

**Comment by netnem** [comment, marked as answer]
````
docker run -d \
  --name=wg-easy \
  -e WG_HOST=wg.example.com \
  -e PASSWORD=mysecure_password \
  -e WG_DEFAULT_DNS=1.1.1.1 \
  -v ~/.wg-easy:/etc/wireguard \
  -p 51820:51820/udp \
  -p 51821:51821/tcp \
  --cap-add=NET_ADMIN \
  --cap-add=SYS_MODULE \
  --network-host \
  --restart unless-stopped \
  weejewel/wg-easy
````
Then inside your host VM add the following - 

#enp6s18 is uplink to your router - check yours with "ip route"  and look for "default via {ip} dev {interface}"  -- {interface} is what you want
#wg0 - wireguard network
```
sudo iptables -t nat -A POSTROUTING -o enp6s18 -j MASQUERADE
sudo iptables -A FORWARD -i enp6s18 -o wg0 -m state --state RELATED,ESTABLISHED -j ACCEPT
sudo iptables -A FORWARD -i wg0 -o enp6s18 -j ACCEPT

#this is useful if you are trying to connect from your LAN to a device on your LAN via the VPN address (OPTIONAL)
sudo iptables -t nat -A POSTROUTING -o wg0 -j MASQUERADE 
```

#make it persistant across reboots
`sudo apt-get install iptables-persistent`

 vi /etc/sysctl.conf
uncomment [net.ipv4.ip_forward=1](url)
sysctl -p 
(or reboot)

Hopefully that gets you going in the right direction - confirmed it works for me on ubuntu 20.04 VM.

**Reply by Mladia** [reply]
This worked for me. In my case ufw was blocking access from the docker network to the services on the host. (Docker traffic is coming from a different subnet)

**Reply by meesteridle** [reply]
This worked for me, too.
